File: gameSrc/audio.js
"use strict";

var constants = require("./constants");
var features = require("./systemFeatures");
var storage = require("./storage");
var assert = require("./functions").assert;

var MUSIC_KEY = "MUSIC_";

// WebAudio context object.
//
var audioContext;

// Music audio node.
//
var musicNode = null;

// SFX audio node.
//
var sfxNode = null;

// if audio sfx and music is deactivated then do not initialize the audio context
if (features.audioSFX || features.audioMusic) {

  // construct new context
  if (window.AudioContext) {
    audioContext = new window.AudioContext();
  } else if (window.webkitAudioContext) {
    audioContext = new window.webkitAudioContext();
  }

  var createNode = (audioContext.createGainNode ? audioContext.createGainNode: audioContext.createGain);

  // construct audio nodes
  if (audioContext) {
    sfxNode = createNode.call(audioContext);
    sfxNode.gain.value = 1;
    sfxNode.connect(audioContext.destination);
    musicNode = createNode.call(audioContext);
    musicNode.gain.value = 0.5;
    musicNode.connect(audioContext.destination);
  }
}

var playSoundOnGainNode = function (gainNode, buffer, loop) {
  var source = audioContext.createBufferSource();

  // is loop enabled ?
  if (loop) {
    source.loop = true;
  }

  source.buffer = buffer;
  source.connect(gainNode);
  source.start(0);

  return source;
};

var musicLoadCb = function (obj) {
  if (constants.DEBUG) assert(obj.value);

  currentMusic.connector = playSoundOnGainNode(musicNode, Base64Helper.decodeBuffer(obj.value), true);
  currentMusic.inLoadProcess = false;
};

// Cache for audio buffers.
//
var buffer = {};

// Current played music object.
//
var currentMusic = {
  inLoadProcess: false,
  connector: null,
  id: null
};

exports.decodeAudio = function (arrayBuffer, successCb, errorCb) {
  audioContext.decodeAudioData(arrayBuffer,successCb,errorCb);
}

//
// Returns the value of the sfx audio node.
//
exports.getSfxVolume = function () {
  if (!audioContext) return;
  return sfxNode.gain.value;
};

//
// Returns the value of the music audio node.
//
exports.getMusicVolume = function () {
  if (!audioContext) return;
  return musicNode.gain.value;
};

//
// Sets the value of the sfx audio node.
//
exports.setSfxVolume = function (vol) {
  if (!audioContext) return;

  if (vol < 0) {
    vol = 0;
  } else if (vol > 1) {
    vol = 1;
  }

  sfxNode.gain.value = vol;
};

//
// Sets the value of the music audio node.
//
exports.setMusicVolume = function (vol) {
  if (!audioContext) return;

  if (vol < 0) {
    vol = 0;
  } else if (vol > 1) {
    vol = 1;
  }

  musicNode.gain.value = vol;
};

//
// Registers an audio buffer object.
//
// @param id
// @param buff
//
exports.registerAudioBuffer = function (id, buff) {
  if (constants.DEBUG) assert(!this.isBuffered(id));

  buffer[id] = buff;
};

//
// Removes a buffer from the cache.
//
exports.unloadBuffer = function (id) {
  if (constants.DEBUG) assert(this.isBuffered(id));

  delete buffer[id];
};

//
//
// @param id
// @return {boolean}
//
exports.isBuffered = function (id) {
  return buffer.hasOwnProperty(id);
};

//
// Plays an empty sound buffer. Useful to
// initialize the audio system.
//
exports.playNullSound = function () {
  if (!audioContext) return;

  var buffer = audioContext.createBuffer(1, 1, 22050);
  var source = audioContext.createBufferSource();

  source.buffer = buffer;
  source.connect(audioContext.destination);
  source.start(0);
};

//
// Plays a sound.
//
// @param {string} id id of the sound that will be played
// @param {boolean=} loop (default:false)
// @return {*}
//
exports.playSound = function (id, loop) {
  if (!audioContext) return null;
  if (constants.DEBUG) assert(buffer[id]);

  return playSoundOnGainNode(sfxNode, buffer[id], loop);
};

//
// Plays a background music.
//
// @param id id of the music object
//
exports.playMusic = function (id) {
  if (!audioContext || currentMusic.inLoadProcess) return false;

  // already playing this music ?
  if (currentMusic.id === id) {
    return;
  }

  // stop old music
  if (currentMusic.connector) {
    this.stopMusic();
  }

  // set meta data
  currentMusic.inLoadProcess = true;
  currentMusic.id = id;
  storage.get(MUSIC_KEY + id, musicLoadCb);

  return true;
};

//
// Stop existing background music.
//
exports.stopMusic = function () {
  if (!audioContext || currentMusic.inLoadProcess) return false;

  // disable current music
  if (currentMusic) {
    currentMusic.connector.noteOff(0);
    currentMusic.connector.disconnect(0);
  }

  // remove meta data
  currentMusic.connector = null;
  currentMusic.id = null;
  currentMusic.inLoadProcess = false;

  return true;
};

File: gameSrc/net/wolfTec/utility/Audio.java
package net.wolfTec.utility;

import net.wolfTec.CustomWarsTactics;
import net.wolfTec.bridges.Globals;
import net.wolfTec.bridges.webAudio.*;
import net.wolfTec.utility.Storage.StorageEntry;

import org.stjs.javascript.*;
import org.stjs.javascript.functions.Callback1;

public class Audio {

	public static final String LOG_HEADER = "AUDIO";

	public static final String MUSIC_KEY = "MUSIC_";

	public static final float DEFAULT_SFX_VOL = 1;
	public static final float DEFAULT_MUSIC_VOL = 0.5f;

	private int apiStatus;

	/**
	 * SFX audio node.
	 */
	private AudioGainNode _sfxNode;

	/**
	 * Music audio node.
	 */
	private AudioGainNode _musicNode;

	/**
	 * WebAudio context object.
	 */
	private AudioContext _context;

	/**
	 * Cache for audio buffers.
	 */
	private Map<String, AudioBuffer> _buffer;

	// current music
	private boolean musicInLoadProcess = false;
	private AudioBufferSource musicConnector = null;
	private String musicID = null;

	/**
     *
     */
	private Callback1<StorageEntry> _musicLoadCallback = new Callback1<Storage.StorageEntry>() {
		@Override public void $invoke(Storage.StorageEntry entry) {
			musicConnector = _playSoundOnGainNode(_musicNode, Globals.Base64Helper.decodeBuffer(entry.value), true);
			musicInLoadProcess = false;
		}
	};

	public void init() {
		try {

			// context
			if (JSObjectAdapter.hasOwnProperty(Global.window, "AudioContext")) {
				_context = new AudioContext();
			} else if (JSObjectAdapter.hasOwnProperty(Global.window, "webkitAudioContext")) {
				_context = new webkitAudioContext();
			} else
				JSGlobal.stjs.exception("noWebKitFound");

			// audio nodes
			_sfxNode = JSObjectAdapter.hasOwnProperty(_context, "createGain") ? _context.createGain() : _context.createGainNode();
			_sfxNode.gain.value = DEFAULT_SFX_VOL;
			_sfxNode.connect(_context.destination);

			_musicNode = JSObjectAdapter.hasOwnProperty(_context, "createGain") ? _context.createGain() : _context.createGainNode();
			_musicNode.gain.value = DEFAULT_MUSIC_VOL;
			_musicNode.connect(_context.destination);

			_buffer = JSCollections.$map();

			Debug.logInfo(LOG_HEADER, "Initialized");

		} catch (Exception e) {
			Debug.logInfo(LOG_HEADER, "Disabled => No usable webAudio API found");
			Features features = CustomWarsTactics.getBean("features");
			features.audioSFX = false;
			features.audioMusic = false;
		}
	}

	private AudioBufferSource _playSoundOnGainNode(AudioGainNode gainNode, AudioBuffer buffer, boolean loop) {
		AudioBufferSource source = _context.createBufferSource();
		source.loop = loop;
		source.buffer = buffer;
		source.connect(gainNode);
		if (apiStatus == 0)
			apiStatus = JSObjectAdapter.hasOwnProperty(source, "start") ? 1 : 2;
		if (apiStatus == 1)
			source.start(0);
		else
			source.noteOn(0);
		return source;
	}

	/** */
	public void decodeAudio(Object arrayBuffer, Callback1<AudioBuffer> successCb, Callback1<Object> errorCb) {
		_context.decodeAudioData(arrayBuffer, successCb, errorCb);
	}

	/**
	 * Returns the value of the sfx audio node.
	 */
	public float getSfxVolume() {
		return (_context == null) ? -1 : _sfxNode.gain.value;
	}

	/**
	 * Returns the value of the music audio node.
	 */
	public float getMusicVolume() {
		return (_context == null) ? -1 : _musicNode.gain.value;
	}

	/**
	 *
	 * @param node
	 * @param volume
	 */
	private void _setVolume(AudioGainNode node, float volume) {
		if (_context == null)
			return;

		if (volume < 0)
			volume = 0;
		else if (volume > 1)
			volume = 1;

		node.gain.value = volume;
	}

	/**
	 * Sets the value of the sfx audio node.
	 *
	 * @param vol
	 */
	public void setSfxVolume(float vol) {
		_setVolume(_sfxNode, vol);
	}

	/**
	 * Sets the value of the music audio node.
	 *
	 * @param vol
	 */
	public void setMusicVolume(float vol) {
		_setVolume(_musicNode, vol);
	}

	/**
	 * Registers an audio buffer object.
	 */
	public void registerAudioBuffer(String id, AudioBuffer buffer) {
		if (isBuffered(id))
			Debug.logCritical(LOG_HEADER, "AlreadyRegistered");
		_buffer.$put(id, buffer);
	}

	/**
	 * Removes a buffer from the cache.
	 */
	public void unloadBuffer(String id) {
		if (isBuffered(id))
			Debug.logCritical(LOG_HEADER, "NotRegistered");
		_buffer.$delete(id);
	}

	/** */
	public boolean isBuffered(String id) {
		return JSObjectAdapter.hasOwnProperty(_buffer, id);
	}

	/**
	 * Plays an empty sound buffer. Useful to initialize the audio system.
	 */
	public void playNullSound() {
		if (_context == null)
			return;
		_playSoundOnGainNode(_sfxNode, _context.createBuffer(1, 1, 22050), false);
	}

	/**
	 * Plays a sound and returns the signal.
	 */
	public AudioBufferSource playSound(String id, boolean loop) {
		if (_context == null)
			return null;
		return _playSoundOnGainNode(_sfxNode, _buffer.$get(id), loop);
	}

	/**
	 * Plays a background music.
	 */
	public boolean playMusic(String id) {
		if (_context == null || musicInLoadProcess)
			return false;

		// already playing this music ?
		if (musicID == id)
			return false;

		if (musicConnector != null)
			stopMusic();

		// set meta data
		musicInLoadProcess = true;
		musicID = id;
		Storage.get(MUSIC_KEY + id, _musicLoadCallback);

		return true;
	}

	/**
	 * Stop existing background music.
	 */
	public boolean stopMusic() {
		if (_context == null || musicInLoadProcess)
			return false;

		// disable current music
		if (musicConnector != null) {
			// api status will be available here
			if (apiStatus == 1)
				musicConnector.stop(0);
			else
				musicConnector.noteOff(0);
			musicConnector.disconnect(0);
		}

		// remove meta data
		musicConnector = null;
		musicID = null;
		musicInLoadProcess = false;

		return true;
	}

}
